File: backend/tools/tools.py
from typing import Type, List, TypedDict, Dict, Any, Optional
from pydantic import BaseModel, Field
from dotenv import load_dotenv
import requests
import sys
import os

# 导入 Backlog 类型用于注解
try:
    from core.memory import Backlog
    from tools.score_management.services import StudentScoreService
    from tools.task.task_organizer import TaskOrganizer
except ImportError:
    try:
        from ..core.memory import Backlog
        from .score_management.services import StudentScoreService
        from .task.task_organizer import TaskOrganizer
    except ImportError:
        Backlog = Any
        StudentScoreService = Any
        TaskOrganizer = Any

load_dotenv()

# 1. 定义工具字典结构
class CustomToolDict(TypedDict):
    type: str
    name: str
    description: str
    parameters: Dict[str, Any]

# 2. 定义 Pydantic 模型（放在使用它们类之前）

class Get_Local_Backlog(BaseModel):
    """获取当前的历史对话记录"""
    pass

class Get_Weather(BaseModel):
    """获取指定地区的实时天气信息"""
    adcode: str = Field(..., description="城市名称或中国城市编码（如 '广州' 或 '440100'）", alias="city")

    class Config:
        populate_by_name = True

class Get_Traffic(BaseModel):
    """获取两点间驾车路况（粗略估计）"""
    origin: str = Field(..., description="起点坐标 'lng,lat'")
    destination: str = Field(..., description="终点坐标 'lng,lat'")
    strategy: int = Field(0, description="路径策略（0为速度优先）")

class Load_Backlog(BaseModel):
    """加载指定日期的对话记录"""
    target_date: str = Field(..., description="目标日期")

class TaskOrganizerTool(BaseModel):
    """生成格式化的待办清单"""
    tasks: List[str] = Field(..., description="任务列表")

class Image_Recognition(BaseModel):
    """多场景图像识别"""
    image_path: str = Field(..., description="图片完整路径")
    scene: str = Field("car", description="识别场景：car/dish/animal/plant/logo/object/ingredient")

class Qwen_WebSearch(BaseModel):
    """通义千问联网搜索问答"""
    query: str = Field(..., description="用户要搜索或提问的问题")

class Locate_IP(BaseModel):
    """根据IP地址获取地理位置信息（基于高德地图API）"""
    ip: str = Field("", description="要查询的IP地址，不传则自动查询当前设备公网IP所在的位置")

class Query_Score(BaseModel):
    """查询学生成绩"""
    student_id: Optional[str] = Field(None, description="学生ID")
    name: Optional[str] = Field(None, description="学生姓名，支持模糊搜索")

class Add_Score(BaseModel):
    """录入或更新学生成绩"""
    student_id: str = Field(..., description="学生ID")
    name: str = Field(..., description="学生姓名")
    scores: Dict[str, float] = Field(..., description="成绩字典，例如 {'数学': 95, '英语': 88}")

class Delete_Score(BaseModel):
    """删除学生信息"""
    student_id: Optional[str] = Field(None, description="学生ID")
    name: Optional[str] = Field(None, description="学生姓名")

import re

def build_tools_list(models: List[Type[BaseModel]]) -> List[CustomToolDict]:
    tool_list: List[CustomToolDict] = []
    for model in models:
        schema = model.model_json_schema()
        tool_description = schema.pop("description", None) or model.__doc__ or ""
        schema.pop("title", None)
        
        # 将类名转换为下划线命名法 (Snake Case)
        # 如 TaskOrganizerTool -> task_organizer_tool
        name = model.__name__
        name = re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()
        
        tool_list.append({
            "type": "function",
            "name": name,
            "description": tool_description,
            "parameters": schema
        })
    return tool_list

class AgentTools:
    
    def __init__(self):
        self.tool_list = build_tools_list([
            Get_Local_Backlog, Get_Weather, Get_Traffic, Load_Backlog, 
            Image_Recognition,
            TaskOrganizerTool, Qwen_WebSearch, Locate_IP,
            Query_Score, Add_Score, Delete_Score
        ])
        self.score_service = StudentScoreService()
        self.task_organizer_service = TaskOrganizer(self)
    
    def get_local_backlog(self, backlog: Backlog):
        """获取对话记录"""
        print(backlog.get_text())

    def get_weather(self, adcode: str = "", **kwargs):
        """获取天气信息"""
        gaode_api_key = os.getenv("Gaode_API_Key")
        target_city = adcode or kwargs.get("city") or kwargs.get("adcode")
        if not target_city:
            print("没有提供城市信息，无法获取天气。")
            return None
        url = f"https://restapi.amap.com/v3/weather/weatherInfo?city={target_city}&key={gaode_api_key}"
        result = requests.get(url, timeout=10).json()
        return result

    def get_traffic(self, origin: str, destination: str, strategy: int = 0):
        """通过高德驾车路径规划 API 粗略估计路况。"""
        gaode_api_key = os.getenv("Gaode_API_Key")
        if not gaode_api_key:
            raise Exception("❌ 环境变量缺失：Gaode_API_Key")
        if not origin or not destination:
            raise ValueError("origin/destination 不能为空，格式应为 'lng,lat'")

        url = "https://restapi.amap.com/v3/direction/driving"
        params = {
            "origin": origin,
            "destination": destination,
            "strategy": str(strategy),
            "extensions": "all",
            "key": gaode_api_key,
        }
        raw = requests.get(url, params=params, timeout=10).json()

        def safe_int(v):
            try:
                return int(float(v))
            except Exception:
                return None

        duration_sec = None
        tmcs_status_counts: Dict[str, int] = {}
        traffic_level = "unknown"

        route = (raw or {}).get("route") or {}
        paths = route.get("paths") or []
        if paths:
            duration_sec = safe_int((paths[0] or {}).get("duration"))
            steps = (paths[0] or {}).get("steps") or []
            for st in steps:
                for t in (st or {}).get("tmcs") or []:
                    status = str((t or {}).get("status") or "").strip()
                    if not status:
                        continue
                    tmcs_status_counts[status] = tmcs_status_counts.get(status, 0) + 1

        if tmcs_status_counts:
            bad = tmcs_status_counts.get("严重拥堵", 0) + tmcs_status_counts.get("拥堵", 0)
            mid = tmcs_status_counts.get("缓行", 0)
            good = tmcs_status_counts.get("畅通", 0)
            if bad > 0:
                traffic_level = "congested"
            elif mid > 0:
                traffic_level = "slow"
            elif good > 0:
                traffic_level = "good"

        return {
            "traffic_level": traffic_level,
            "duration_sec": duration_sec,
            "tmcs_status_counts": tmcs_status_counts if tmcs_status_counts else None,
            "raw": raw,
        }

    def load_backlog(self, backlog: Backlog, target_date: str):
        """加载指定日期的对话记录"""
        return backlog.load_backlog(target_date)

    def get_resource_path(self, relative_path):
        """智能获取资源路径"""
        if hasattr(sys, '_MEIPASS'):
            base_path = getattr(sys, '_MEIPASS', os.path.abspath("."))
            return os.path.join(base_path, relative_path)
        
        dev_path = os.path.join(os.path.abspath("."), relative_path)
        return dev_path

    def image_recognition(self, image_path: str, scene: str = "car"):
        """多场景图像识别"""
        import base64
        baidu_api_key = os.getenv("BAIDU_API_KEY")
        baidu_secret_key = os.getenv("BAIDU_SECRET_KEY")

        if not all([baidu_api_key, baidu_secret_key]):
            raise Exception("❌ 百度API配置缺失")

        def get_baidu_token():
            token_url = f"https://aip.baidubce.com/oauth/2.0/token?grant_type=client_credentials&client_id={baidu_api_key}&client_secret={baidu_secret_key}"
            res_json = requests.post(token_url, timeout=10).json()
            return res_json.get("access_token")

        if not os.path.exists(image_path):
            # 这里的 f-string 可能在某些旧版分析器报错，尝试简化
            msg = "文件不存在: " + str(image_path)
            raise FileNotFoundError(msg)

        with open(image_path, "rb") as f:
            img_b64 = base64.b64encode(f.read()).decode("utf-8")

        token = get_baidu_token()
        api_urls = {
            "car": "https://aip.baidubce.com/rest/2.0/image-classify/v1/car",
            "animal": "https://aip.baidubce.com/rest/2.0/image-classify/v1/animal",
            "plant": "https://aip.baidubce.com/rest/2.0/image-classify/v1/plant",
            "object": "https://aip.baidubce.com/rest/2.0/image-classify/v1/object_detect",
            "logo": "https://aip.baidubce.com/rest/2.0/image-classify/v2/logo",
            "dish": "https://aip.baidubce.com/rest/2.0/image-classify/v2/dish",
            "ingredient": "https://aip.baidubce.com/rest/2.0/image-classify/v1/ingredient"
        }

        if scene not in api_urls:
            return None

        url = f"{api_urls[scene]}?access_token={token}"
        data = {"image": img_b64}
        baidu_result = requests.post(url, data=data, timeout=10).json()

        if "result" in baidu_result and len(baidu_result["result"]) > 0:
            return baidu_result["result"][0]
        return None
        
    def qwen_websearch(self, query: str):
        """通义千问联网搜索问答"""
        api_key = os.getenv("DASHSCOPE_API_KEY")
        url = "https://dashscope.aliyuncs.com/api/v1/services/aigc/text-generation/generation"
        headers = {"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"}
        payload = {
            "model": "qwen-flash",
            "input": {"messages": [{"role": "user", "content": query}]},
            "parameters": {"enable_search": True}
        }
        try:
            resp = requests.post(url, headers=headers, json=payload)
            return resp.json()['output']['text'].strip()
        except Exception as e:
            return f"搜索失败：{str(e)}"
    
    def locate_ip(self, ip: str = ""):
        """根据IP地址获取地理位置信息（基于高德地图API）"""
        gaode_api_key = os.getenv("Gaode_API_Key")
        if not gaode_api_key:
            raise Exception("❌ 环境变量缺失：Gaode_API_Key")
        
        url = "https://restapi.amap.com/v3/ip"
        params = {"key": gaode_api_key}
        if ip:
            params["ip"] = ip
        
        result = requests.get(url, params=params, timeout=10).json()
        return result

    def task_organizer(self, tasks: List[str]):
        """生成格式化的待办清单"""
        tasks_text = "\n".join(tasks)
        # 默认使用广州的 adcode 进行天气查询，实际应用中可动态获取
        return self.task_organizer_service.generate_today_itinerary(tasks_text, city_adcode="440100")
    

    def query_score(self, student_id: str = None, name: str = None):
        """查询学生成绩"""
        return self.score_service.query_students(student_id=student_id, name=name)

    def add_score(self, student_id: str, name: str, scores: Dict[str, float]):
        """录入或更新成绩"""
        return self.score_service.add_score(student_id, name, scores)

    def delete_score(self, student_id: str = None, name: str = None):
        """删除学生信息"""
        success = self.score_service.delete_student(student_id=student_id, name=name)
        return "删除成功" if success else "未找到对应学生或删除失败"

File: backend_kotlin/src/main/kotlin/com/aegis/backend/tools/Tools.kt
package com.aegis.backend.tools

import com.aegis.backend.core.Backlog
import com.aegis.backend.core.ChatHistoryResult
import com.aegis.backend.core.EnvConfig
import com.aegis.backend.tools.precise_search.PreciseSearch
import com.aegis.backend.tools.score_management.StudentScoreService
import kotlinx.serialization.json.JsonElement
import com.aegis.backend.tools.task.TaskOrganizer
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.util.concurrent.TimeUnit

/**
 * Agent 工具集 - 封装所有可调用工具
 */
class AgentTools {
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    val scoreService = StudentScoreService()
    val taskOrganizerService = TaskOrganizer(this)
    private val preciseSearch = PreciseSearch()

    /**
     * 获取天气信息
     */
    /**
     * 通过高德地理编码 API 将中文地名解析为 adcode（区域编码）
     * 支持省/市/区级单位，例如"汕头"→440500、"濠江区"→440512
     * 高德天气 API 要求使用 adcode，不支持中文城市名直接查询
     */
    private fun getAdcodeByCityName(cityName: String): String? {
        val apiKey = EnvConfig.gaodeApiKey
        if (apiKey.isBlank()) return null
        // 不传 city 参数，让高德自动匹配省市区各级
        val url = "https://restapi.amap.com/v3/geocode/geo?address=$cityName&key=$apiKey"
        return try {
            val request = Request.Builder().url(url).get().build()
            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return null
            val json = JSONObject(body)
            if (json.optString("status") != "1") {
                println(">>> 高德地理编码API返回失败: ${json.optString("info")} (city=$cityName)")
                return null
            }
            val geocodes = json.optJSONArray("geocodes")
            if (geocodes == null || geocodes.length() == 0) {
                println(">>> 未找到 $cityName 的 adcode")
                return null
            }
            // 取第一个匹配结果的 adcode（区域编码），支持省/市/区各级
            val adcode = geocodes.getJSONObject(0).optString("adcode").ifBlank { null }
            if (adcode != null) {
                println(">>> 地理编码: 「$cityName」→ adcode=$adcode")
            }
            adcode
        } catch (e: Exception) {
            println(">>> 获取adcode失败: ${e.message}")
            null
        }
    }

    fun getWeather(adcode: String = ""): Any? {
        var targetCity = adcode.ifBlank { return null }
        val apiKey = EnvConfig.gaodeApiKey
        if (apiKey.isBlank()) {
            println("没有配置 Gaode_API_Key")
            return null
        }
        // 如果传入了中文城市名（非纯数字），先通过地理编码转换为 adcode
        if (targetCity.any { it in '0'..'9' }.not()) {
            val resolved = getAdcodeByCityName(targetCity)
            if (resolved != null) {
                println(">>> 城市名「$targetCity」→ adcode: $resolved")
                targetCity = resolved
            } else {
                println(">>> 警告：无法解析城市名「$targetCity」，直接传参尝试")
            }
        }
        val url = "https://restapi.amap.com/v3/weather/weatherInfo?city=$targetCity&key=$apiKey"
        return try {
            val request = Request.Builder().url(url).get().build()
            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return null
            val json = JSONObject(body)
            // 检查高德 API 返回状态：0=失败
            val status = json.optString("status", "0")
            val info = json.optString("info", "")
            if (status != "1") {
                println(">>> 高德天气API返回失败: $info (city=$targetCity)")
                return null
            }
            json.toDeepMap()
        } catch (e: Exception) {
            println("获取天气失败: ${e.message}")
            null
        }
    }

    /**
     * 获取路况信息
     */
    fun getTraffic(origin: String, destination: String, strategy: Int = 0): Map<String, Any?>? {
        val apiKey = EnvConfig.gaodeApiKey
        if (apiKey.isBlank()) {
            println("❌ 环境变量缺失：Gaode_API_Key")
            return null
        }
        if (origin.isBlank() || destination.isBlank()) return null

        val url = "https://restapi.amap.com/v3/direction/driving" +
                "?origin=$origin&destination=$destination&strategy=$strategy&extensions=all&key=$apiKey"

        return try {
            val request = Request.Builder().url(url).get().build()
            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return null
            val raw = JSONObject(body)

            var durationSec: Long? = null
            val tmcsStatusCounts = mutableMapOf<String, Int>()
            var trafficLevel = "unknown"

            val route = raw.optJSONObject("route")
            val paths = route?.optJSONArray("paths")
            if (paths != null && paths.length() > 0) {
                val path = paths.getJSONObject(0)
                durationSec = path.optLong("duration", -1).takeIf { it >= 0 }
                val steps = path.optJSONArray("steps")
                if (steps != null) {
                    for (i in 0 until steps.length()) {
                        val step = steps.getJSONObject(i)
                        val tmcs = step.optJSONArray("tmcs")
                        if (tmcs != null) {
                            for (j in 0 until tmcs.length()) {
                                val status = tmcs.getJSONObject(j).optString("status", "").trim()
                                if (status.isNotBlank()) {
                                    tmcsStatusCounts[status] = tmcsStatusCounts.getOrDefault(status, 0) + 1
                                }
                            }
                        }
                    }
                }
            }

            val bad = (tmcsStatusCounts["严重拥堵"] ?: 0) + (tmcsStatusCounts["拥堵"] ?: 0)
            val mid = tmcsStatusCounts["缓行"] ?: 0
            val good = tmcsStatusCounts["畅通"] ?: 0
            trafficLevel = when {
                bad > 0 -> "congested"
                mid > 0 -> "slow"
                good > 0 -> "good"
                else -> "unknown"
            }

            mapOf(
                "traffic_level" to trafficLevel,
                "duration_sec" to durationSec,
                "tmcs_status_counts" to tmcsStatusCounts,
                "raw" to raw.toDeepMap()
            )
        } catch (e: Exception) {
            println("获取路况失败: ${e.message}")
            null
        }
    }

    /**
     * 联网搜索 — 优先使用 WEB_SEARCH_CONFIG 配置，回退旧版 DashScope 逻辑
     */
    fun webSearch(query: String): String {
        val cfg = EnvConfig.webSearchConfig
        val apiKey = cfg.apiKey.ifBlank {
            EnvConfig.dashscopeApiKey.ifBlank { EnvConfig.openaiApiKey }
        }
        if (apiKey.isBlank()) return "请先在 config.json 中配置 WEB_SEARCH_CONFIG 或 DASHSCOPE_API_KEY"

        val url = cfg.baseUrl.ifBlank {
            "https://dashscope.aliyuncs.com/api/v1/services/aigc/text-generation/generation"
        }
        val model = cfg.model.ifBlank { "qwen3.5-flash" }

        val jsonBody = JSONObject().apply {
            put("model", model)
            put("input", JSONObject().apply {
                put("messages", listOf(
                    mapOf("role" to "user", "content" to query)
                ))
            })
            put("parameters", JSONObject().apply {
                put("enable_search", true)
            })
        }

        return try {
            val request = Request.Builder()
                .url(url)
                .addHeader("Authorization", "Bearer $apiKey")
                .addHeader("Content-Type", "application/json")
                .post(jsonBody.toString().toRequestBody("application/json".toMediaType()))
                .build()
            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return "搜索失败：无响应"
            val json = JSONObject(body)
            // 先检查 API 是否返回了错误
            if (json.has("code")) {
                val code = json.optString("code", "")
                val msg = json.optString("message", "")
                return "搜索失败：[$code] $msg"
            }
            val text = json.optJSONObject("output")?.optString("text", "")?.trim()
            if (text.isNullOrEmpty()) {
                return "搜索失败：API 返回异常\n$body"
            }
            text
        } catch (e: Exception) {
            "搜索失败：${e.message}"
        }
    }

    /**
     * 通义千问联网搜索（旧名兼容，委托到 webSearch）
     */
    fun qwenWebsearch(query: String): String = webSearch(query)

    /**
     * 精准搜索 — 接收关键词，AI 扩展联想词后在 backlog JSON 文件中匹配
     * @param keyword 搜索关键词
     * @return 格式化的搜索结果文本
     */
    fun preciseSearch(keyword: String): String {
        return preciseSearch.search(keyword)
    }

    /**
     * 查询学生成绩
     */
    fun queryScore(studentId: String? = null, name: String? = null): List<Map<String, Any>> {
        return scoreService.queryStudents(studentId = studentId, name = name).map { it ->
            mapOf(
                "student_id" to it.student_id,
                "name" to it.name,
                "scores" to it.scores
            )
        }
    }

    /**
     * 录入/更新成绩
     */
    fun addScore(studentId: String, name: String, scores: Map<String, JsonElement>): String {
        return scoreService.addScore(studentId, name, scores)
    }

    /**
     * 删除学生
     */
    fun deleteScore(studentId: String? = null, name: String? = null): String {
        val success = scoreService.deleteStudent(studentId = studentId, name = name)
        return if (success) "删除成功" else "未找到对应学生或删除失败"
    }

    /**
     * 获取本地对话记录
     */
    fun getLocalBacklog(backlog: Backlog) {
        println(backlog.getText())
    }

    /**
     * 加载对话记录
     */
    fun loadBacklog(backlog: Backlog, targetDate: String): Map<String, ChatHistoryResult> {
        return backlog.loadBacklog(targetDate)
    }

    /**
     * 根据IP地址获取地理位置信息（基于高德地图API）
     */
    fun locateIp(ip: String = ""): Any? {
        val apiKey = EnvConfig.gaodeApiKey
        if (apiKey.isBlank()) {
            println("❌ 环境变量缺失：Gaode_API_Key")
            return null
        }

        val url = if (ip.isNotBlank()) {
            "https://restapi.amap.com/v3/ip?key=$apiKey&ip=$ip"
        } else {
            "https://restapi.amap.com/v3/ip?key=$apiKey"
        }

        return try {
            val request = Request.Builder().url(url).get().build()
            val response = client.newCall(request).execute()
            val body = response.body?.string()
            if (body != null) JSONObject(body).toDeepMap() else null
        } catch (e: Exception) {
            println("获取IP定位失败: ${e.message}")
            null
        }
    }

    /**
     * 日程规划
     */
    fun taskOrganizer(tasks: List<String>): String {
        val tasksText = tasks.joinToString("\n")
        return taskOrganizerService.generateTodayItinerary(tasksText, cityAdcode = "440100")
    }
}

/**
 * JSONObject 递归转 Map 的辅助扩展（org.json 内置 toMap() 不会递归转换嵌套对象）
 */
fun JSONObject.toDeepMap(): Map<String, Any> {
    val map = mutableMapOf<String, Any>()
    this.keys().forEach { key ->
        val value = this[key]
        map[key] = when (value) {
            is JSONObject -> value.toDeepMap()
            is org.json.JSONArray -> {
                (0 until value.length()).map { i ->
                    val item = value[i]
                    if (item is JSONObject) item.toDeepMap() else item
                }
            }
            else -> value
        }
    }
    return map
}
